### PR TITLE
[FIX] Allow shallow renderer to work with new API

### PR DESCRIPTION
**Pull request by mprobber**
**Before submitting a pull request,** please make sure the following is done:

1. Fork [the repository](https://github.com/facebook/react) and create your branch from `master`.
2. Run `yarn` in the repository root.
3. If you've fixed a bug or added code that should be tested, add tests!
4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
5. Run `yarn test-prod` to test in the production environment. It supports the same options as `yarn test`.
6. If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
9. Run the [Flow](https://flowtype.org/) typechecks (`yarn flow`).
10. If you haven't already, complete the CLA.

**Learn more about contributing:** https://reactjs.org/docs/how-to-contribute.html

Hi,
I really like the new `contextType` API in 16.6, but the shallow renderer doesn't seem to provide support for it.  This pull request allows you to pass in an object for `context`, and if `contextType` is defined, will shallow render the component with the provided context, as it did for the old context API.

### CODE REVIEW

File: packages/react-test-renderer/src/ReactShallowRenderer.js [hunk 1]
```diff
@@ -143,7 +143,12 @@ class ReactShallowRenderer {
 
     this._rendering = true;
     this._element = element;
-    this._context = getMaskedContext(element.type.contextTypes, context);
+
+    if (element.type.contextType) {
```
Comment (gaearon) [inline packages/react-test-renderer/src/ReactShallowRenderer.js]: This should only be enabled for class components.

File: packages/react-test-renderer/src/ReactShallowRenderer.js [hunk 2]
```diff
@@ -143,7 +143,12 @@ class ReactShallowRenderer {
...
+      this._context = context;
```
Comment (gaearon) [inline packages/react-test-renderer/src/ReactShallowRenderer.js]: This is the legacy context object. It has no relation to the new context API you're implementing.

**Comment by mprobber** [comment]
@gaearon

I agree with you that the shallow renderer should use the value of the default context if no context is provided, but I'm not really sure about rendering two levels deep and not being transparent about it in the shallow renderer for a select few components.

Looking at https://github.com/facebook/react/blob/c954efa70f44a44be9c33c60c57f87bea6f40a10/packages/react-reconciler/src/ReactFiberClassComponent.js#L362 there is no way for a component to support both the `contextType` and the legacy `contextTypes` APIs, so it doesn't seem like it would be problematic to use the same argument to set context on a component (just an object, used to set this.context).   I don't have too much context (*ba-dum-tss*) about implementation details though, so I'm curious if there's something I'm missing.  I'll update the PR so the default context renders if none is provided.

**Comment by jfirebaugh** [comment]
> It doesn't seem like it would be problematic to use the same argument to set context on a component (just an object, used to set this.context).

Coming here from https://github.com/airbnb/enzyme/issues/1553, via https://stackoverflow.com/questions/53711166/unable-to-set-context-in-an-enzyme-test, and that's pretty much exactly how I expected it to work.

**Comment by minznerjosh** [comment]
To @gaearon's point, using the second argument to support `contextType` would create some funkiness in enzyme.

It seems reasonable at first because, with the plain react shallow renderer, you're only ever rendering one component at a time, so you have a lot of control over what context you pass when you render a component; you _know_ not to pass legacy context if the component you're testing is using `contextType`.

But Enzyme has the concept of `dive()`ing, allowing you to render components further down in the tree. And, most importantly, when you `dive()` a component, the **legacy** context of any providers further up in the tree will be passed as context (via the second argument to `.render()`).

So, if this change were to be released, and no changes in Enzyme were made, if you were to `dive()` a component that uses `contextType` you would end up with a bunch of **legacy** context as `this.context` in your component.

Now, of course, Enzyme could be updated to inspect the component being `dive()`d to determine whether new or legacy context should be passed. But I still think it's worth calling out that this change is a little more "break-y" then it appears at first. And I think this potentially odd behavior is a decent smell for the fact that new context and legacy context are actually totally different APIs, and we maybe shouldn't overload the second arg to `.render()` to handle both.

**Comment by mprobber** [comment]
@minznerjosh 

The only thing coming to mind to that would avoid overloading the second argument, and supporting `ContextType` would be we passing a Map of `React.Context`, value pairs, but that seems very, very clunky.

I'd think that a follow-up change should be made to enzyme regardless if this PR gets merged.  Currently, if you shallow render a component with `contextType`, `this.context` is the empty object, as opposed to the default value of the context.  So, while this would make `dive()` behave bizarrely for components that use `contextType`, they already behaves incorrectly under enzyme.  Regardless, I'd be happy to craft a pull request for enzyme when any changes are made.

Let me know if you come up with an API that doesn't overload the second argument to render that's not super clunky.  I'd be happy to implement it.

**Comment by gaearon** [comment]
>But Enzyme has the concept of dive()ing

Oh my. This is the first time I'm hearing and reading about this API 😳This seems like a real can of wormsn.

**Comment by ljharb** [comment]
`.dive()` is just sugar for a feature enzyme has always had since v1, the ability to find an arbitrary custom component and create a new shallow wrapper from that point.

**Comment by gaearon** [comment]
The confusing thing to me is that the updates in the parent component wouldn't "reflect" in that "dived" version for `shallow`. But they would for `mount`. Unless I'm missing something.

**Comment by minznerjosh** [comment]
@gaearon I'm pretty sure  you're correct about this. If you change state/props in the parent, you'd need to re-dive to see the changes reflected in the "dived" version. This, however, probably isn't _super_ surprising to users given the immutable nature of non-root enzyme wrappers.

However, it does present itself if you ever want to test component lifecycle methods in the dived child by manipulating the props/state of the parent.

**Comment by ljharb** [comment]
That's also the case for both shallow and mount for any derived wrapper - ie, any `.find`-produced wrapper is not updated if the parent updates, you always have to re-find from the root.

**Comment by gaearon** [comment]
Thanks for explaining!

**Comment by DavyJohnes** [comment]
Guys, is there any workaround to pass context to component which uses new context API?

**Comment by d4wei** [comment]
@mprobber @gaearon Any chance we might move forward with this, or decide on an alternative API in the near future? Having shallow rendering support the new `contextType` API would be really useful!

If we are thinking about alternatives, would passing a map of Provider/value pairs when creating an instance of `shallowRenderer`(instead of passing it when `render` is called) be a less clunky way to do this while avoiding the problems @minznerjosh described? I think in a certain sense, non-default values for contexts can also be interpreted as "configurations" for a `shallowRenderer` instance.

> Guys, is there any workaround to pass context to component which uses new context API?

@DaveVoyles If the context your component uses **happens to be an object**, you can use the legacy API to test it for now by setting an appropriate value for `contextTypes` on your component at test time. I've documented this workaround in the [Stack Overflow question](https://stackoverflow.com/a/56483320/11610682) that led me here.

**Comment by comp615** [comment]
Wanted to throw my hat into the ring too, having a solution here would greatly help unblocking our upgrade path to remove old-style, legacy context. Right now we code-modded most of it away and the code is functioning, but tests are an absolute mess. Can explore the workaround detailed above, but that feels a bit worse.

**Comment by mjancarik** [comment]
I created module for workaround https://www.npmjs.com/package/shallow-with-context. The module works well in our projects.

**Comment by ljharb** [comment]
bump

**Comment by gaearon** [comment]
The shallow renderer has been extracted so please port this change to https://github.com/NMinhNguyen/react-shallow-renderer if you'd like. Thank you!

**Pull request closed without merging**